src/OrleansRuntime/MultiClusterNetwork/MultiClusterData.cs

```diff
@@ -0,0 +1,287 @@
+                       list.Add(g.Key, g.Value);
+                resultlist = list;
+            }
+
+            MultiClusterConfiguration resultconf;
+            if (exclude.Configuration == null)
+                resultconf = this.Configuration;
+            else
+                resultconf = null;
+
+            return new MultiClusterData(resultlist, resultconf);
+        }
+
+        public bool Equals(MultiClusterData other)
+        {
+            if ((this.Configuration == null) != (other.Configuration == null))
+              return false;
+
+            if (this.Gateways.Count != other.Gateways.Count)
+              return false;
+
+            if ((this.Configuration != null) && !this.Configuration.Equals(other.Configuration))
+              return false;
+
+            foreach (var g in this.Gateways)
+            {
+                GatewayEntry othergateway;
+                if (!other.Gateways.TryGetValue(g.Key, out othergateway))
+                    return false;
+                if (!g.Value.Equals(othergateway))
+                    return false;
+            }
+
+            return true;
+        }
+
+    }
+
+     
+
+    /// <summary>
+    /// Information about gateways, as stored/transmitted in the multicluster network.
+    /// </summary>
+    [Serializable]
+    public class GatewayEntry : IMultiClusterGatewayInfo, IEquatable<GatewayEntry>, IComparable<GatewayEntry>
+    {
+        public string ClusterId { get; set; }
+
+        public SiloAddress SiloAddress { get; set; }
+
+        public GatewayStatus Status { get; set; }
+
+        /// <summary>
+        /// UTC timestamp of this gateway entry.
+        /// </summary>
+        public DateTime HeartbeatTimestamp { get; set; }
+
+
+        /// <summary>
+        /// Whether this entry has expired based on its timestamp.
+        /// </summary>
+        public bool Expired
+        {
+            get
+            {
+                return DateTime.UtcNow - HeartbeatTimestamp > ExpiresAfter;
+            }
+        }
+        /// <summary>
+        /// time after which entries expire.
+        /// </summary>
+        public static TimeSpan ExpiresAfter = new TimeSpan(hours: 0, minutes: 30, seconds: 0);
+
+        public bool Equals(GatewayEntry other)
+        {
+            return SiloAddress.Equals(other.SiloAddress)
+                && Status.Equals(other.Status)
+                && HeartbeatTimestamp.Equals(other.HeartbeatTimestamp)
+                && ClusterId.Equals(other.ClusterId);
+        }
+
+        public override int GetHashCode()
+        {
+            return SiloAddress.GetHashCode() ^ ((int)Status << 8) ^ HeartbeatTimestamp.GetHashCode() ^ ClusterId.GetHashCode();
+        }
+
+        public int CompareTo(GatewayEntry other)
+        {
+            var diff = ClusterId.CompareTo(other.ClusterId);
+            if (diff != 0) return diff;
+            diff = SiloAddress.ToString().CompareTo(other.SiloAddress.ToString());
```

Comment:
Yes, using `SiloAddress.CompareTo` is a bad idea because it only compares generations (I promptly stepped into this trap when first implementing it). Unfortunately, the convenient `IpEndpoint.IpAddress` is deprecated because it is incompatible with IPv6. So for now, I am using IPEndPoint.ToString() (I am calling it only once).

So, I am currently using IpEndpoint.ToString().